### PR TITLE
GraphTheory added with Djikstra O(nlogn) implementation

### PR DESCRIPTION
**Pull request by rrkarim**

### CODE REVIEW

File: GraphTheory/Dijkstra.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+#include <queue>
+
+using namespace std;
+
+
+vector<vector< pair<int, int>>> g; // global graph
+int vertices, edges;
+
+vector<int> dijkstra(const int &start) {
+    const int INF = 1000000; // initially all distances are equal infinity
```
Comment (varunjm) [inline GraphTheory/Dijkstra.cpp]: Using INT_MAX  from the <climits> header is a better way to do this.

Comment (rrkarim) [reply GraphTheory/Dijkstra.cpp]: Ok, but not the INT_MAX but INT_INF.

File: GraphTheory/Dijkstra.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
+#include <iostream>
+#include <vector>
+#include <queue>
+
+#define INF 2147483647
```
Comment (faheel) [inline GraphTheory/Dijkstra.cpp]: A better way is
```C++
const int INF = INT_MAX;
```

File: GraphTheory/Dijkstra.cpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+using namespace std;
+
+
+vector<vector< pair<int, int>>> g; // global graph
+int vertices, edges;
+
```
Comment (faheel) [inline GraphTheory/Dijkstra.cpp]: - typedef `vector<vector<pair<int, int>>>` to `Graph`,
- Rename `g` to `graph`,
- Make `graph`, `vertices` and `edges` as local variables of `main()`. You can pass the graph as a constant reference (`const Graph& graph`) to `dijkstra()`.

File: GraphTheory/Dijkstra.cpp [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	g.resize(vertices + 1);
+
+	cout << "Now enter your edges in format: \"u v w\". Where u and w are vertices and w is the distance between them." << endl;
+	for(int i = 0; i < edges; ++i) {
+		int temp_start, temp_end, temp_distance;
+		cin >> temp_start >> temp_end >> temp_distance;
+		g[temp_start].push_back({temp_end, temp_distance});
+	}
+
+	vector<int> d = dijkstra(1);
```
Comment (faheel) [inline GraphTheory/Dijkstra.cpp]: Rename `d` to `distances` or `min_distances`

File: GraphTheory/Dijkstra.cpp [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+	//cout << "Distances from the first vertix to the rest : " << endl;
+    for(int i = 2; i <= vertices; ++i)
+        cout << "Distance from 1 to " << i << " is equal to : " << d[i] << endl;
+
+}
```
Comment (faheel) [inline GraphTheory/Dijkstra.cpp]: Missing `return` statement

File: GraphTheory/Dijkstra.cpp [hunk 6]
```diff
@@ -0,0 +1,55 @@
+#include <iostream>
+#include <vector>
+#include <queue>
+
```
Comment (faheel) [inline GraphTheory/Dijkstra.cpp]: Add the "description comment" at the very top as mentioned [here](https://github.com/faheel/Algos/blob/master/CONTRIBUTING.md#general-guidelines).

**Review by faheel — CHANGES_REQUESTED** [review]

File: GraphTheory/Dijkstra.cpp (hunk 6 above)
Comment (rrkarim) [reply GraphTheory/Dijkstra.cpp]: Done.

File: GraphTheory/Dijkstra.cpp [hunk 7]
```diff
@@ -1,17 +1,22 @@
...
 using namespace std;
 
+typedef vector<vector< pair<int, int>>> Graph; 
 
-vector<vector< pair<int, int>>> g; // global graph
-int vertices, edges;
-
-vector<int> dijkstra(const int &start) {
-	vector<int> distances(g.size(), INF);
+vector<int> dijkstra(const int &start, const Graph& graph, const int& vertices, const int& edges) {
```
Comment (faheel) [inline GraphTheory/Dijkstra.cpp]: There's no need to have `vertices` and `edges` as arguments

File: GraphTheory/Dijkstra.cpp [hunk 8]
```diff
@@ -35,21 +43,24 @@ vector<int> dijkstra(const int &start) {
...
+	int vertices, edges;
 	cout << "Enter the number of vertices and edges : " << endl;
 	cin >> vertices >> edges;
-
-	g.resize(vertices + 1);
+	
+	Graph graph;
+	graph.resize(vertices + 1);
 
 	cout << "Now enter your edges in format: \"u v w\". Where u and w are vertices and w is the distance between them." << endl;
```
Comment (faheel) [inline GraphTheory/Dijkstra.cpp]: The following will be more comprehensible:
```C++
cout << "Enter the edges in the format:" << "\n";
cout << "a w b" << "\n";
cout << "where \'a\' and \'b\' are vertices, and \'w\' is the distance between them." << "\n";
```

**Review by faheel — CHANGES_REQUESTED** [review]
- Use consistent indentation of four spaces per level
- Add a single space between `for`, `while`, `if` and the following `(`
- Rename to folder to just `Graph`

**Comment by faheel** [comment]
In case you missed them, other changes are still pending. Please read the review above.

**Comment by rrkarim** [comment]
Anything else?

**Comment by faheel** [comment]
Sorry for not reviewing your PR back then. The project has been updated quite a lot, so if you're still interested in getting this PR merged, feel free to re-open it and add a comment stating that you are, and I'll help you in making the required changes. Sorry once again for replying so late, and thanks a lot for taking the time to contribute to the project.

**Pull request closed without merging**